Fix planter nonstop complaining

rivon0507 opened the pull request:
Closes #10985 

# Changes proposed in this pull request
- Similar to #10956 

## Testing
- [x] Yes I tested this before submitting it.
- [ ] I also did a multiplayer test.

Review please

rivon0507 committed:
Fix planter nonstop complaining

Thodor12 commented:
I don't see how this would change anything. `setIdleAtJob` only sets a boolean field on the AI. This doesn't modify any behaviour except for removing 1 extra invocation of the method call.

Please explain why this would fix it.

rivon0507 commented:
Before, the planter is set to `idleAtJob=true` at the start of `pickField()` and only set to `idleAtJob=false` when it finds a field to work on (`fieldToWork` is not `null`), but the planter has the same mechanics (?) as the farmer, working a field once per day. So if `fieldTowork` is `null`, the planter will stay `idleAtJob` for the rest of the day and will complain.

Thodor12 commented on src/main/java/com/minecolonies/core/entity/ai/workers/production/agriculture/EntityAIWorkPlanter.java:
```diff
@@ -142,6 +139,20 @@ private IAIState pickField()
+     * @param module the extensions assigned to the planter's hut
+     * @return false if we have at least one assigned plantation field
+     */
+    public boolean checkForField(@NotNull BuildingExtensionsModule module)
```
In that case I'd probably just move the method calls around and not introduce a new method for it, not worth it.
So remove this method.

Raycoms replied:
this seems the cleanest solution for me tbh

rivon0507 replied:
> In that case I'd probably just move the method calls around and not introduce a new method for it, not worth it. So remove this method.

Done :)

armele commented:
That boolean seems to source the predicate for the dayRollOverPredicate in TimeBasedHappinessModifier.dayEnd(), unless my code trace went astray.   I stopped short of tracing the code path in debug mode since @rivon0507 has got this one. :)

So my theory was that any day that ends with isIdleAtJob = true, when TimeBasedHappinessModifier.dayEnd is called then the days of idleness are incremented (even if work was done earlier in the day).  And reviewing the planter AI supported that theory.  (Again, not tested or traced - just reviewed through code.)

  HappinessRegistry.idleatjobFunction = DEFERRED_REGISTER_HAPPINESS_FUNCTION.register(IDLEATJOB_FUNCTION.getPath(), () -> new HappinessFunctionEntry(data -> data.isIdleAtJob() ? 0.5 : 1.0));

Thodor12 commented:
Ok actually hold on this is not a bug at all. You're saying yourself "if the worker can't find a field to work on", that definition constitutes being idle. So he should be complaining if there's nothing to do.

Rather, figure out why fieldToWorkOn is returning null.

rivon0507 commented:
fieldToWorkOn is returning null when all field have been worked for the day. So yes, the citizen can be idle, but should not complain.

Thodor12 commented:
Other side of the spectrum, if you have a field, but nothing is able to grow due to whatever reason. You do want to hear about it that he's unable to work.

Thodor12 commented:
The proposed solution is too straight forward and inhibits legitimate cases where a complaint should be thrown, so please come u with a different solution.

Raycoms commented:
We've been using idle at job as an unexpected idle at job so far. We'd need a new flag for expected idle

-------- Original Message --------
On 6/30/25 23:43, Thom van den Akker  wrote:

> Thodor12 left a comment [(ldtteam/minecolonies#10986)](https://github.com/ldtteam/minecolonies/pull/10986#issuecomment-3020901536)
>
> The proposed solution is too straight forward and inhibits legitimate cases where a complaint should be thrown, so please come u with a different solution.
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/ldtteam/minecolonies/pull/10986#issuecomment-3020901536), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/ABRD3S24EEY4VLTU647QDTD3GGVPXAVCNFSM6AAAAACAPJNJTWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTAMRQHEYDCNJTGY).
> You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

rivon0507 commented:
> The proposed solution is too straight forward and inhibits legitimate cases where a complaint should be thrown, so please come u with a different solution.

I’d like to understand better — could you list the legitimate cases that might be inhibited by the current solution? That way, I can make sure we’re not accidentally suppressing valid complaints.



> Other side of the spectrum, if you have a field, but nothing is able to grow due to whatever reason. You do want to hear about it that he's unable to work.

That scenario wasn’t handled before either, so I don’t think this fix inhibits it. I agree it’s a valid case, but it seems more like a feature request than a bug regression

armele commented:
If it's helpful, here's what I was thinking: https://github.com/ldtteam/minecolonies/compare/version/main...armele:minecolonies:complaining_planters

Basically the valid condition is the NO_FREE_FIELDS situation. Instead of starting pickField() pessimistically (by setting idle to true) we move that inside the two spots where it returns an early IDLE.  The first I suspect shouldn't be possible , but is included for completeness.  The second should correspond to the situation where nothing has been assigned to the planter.

Raycoms commented:
Would you be willing to adjust the setIdle to "setJobStatus" with an enum that is stuck,idle,working, and we then properly set stuck/working everywhere for now and for the complaints/happiness we check for stuck?

In a future PR we could then also set idle and use this for some nice things like going to the tavern or so =D

uecasm commented:
`STUCK` or `BORED`?  To me `STUCK` implies a pathfinding issue.

MotionlessTrain commented:
In cases where setIdleAtJob is currently used, it is actually STUCK

BORED could be an alternate name for IDLE (nothing to do at the moment, but there is nothing wrong either at this moment)

uecasm commented:
Idk, I would think "idle" means "I did things today but have nothing to do right this moment" (good) while "bored" means "I haven't had anything to do today" (bad).  (I also have a lovely bikeshed.)

rivon0507 committed:
Merge branch 'version/main' into fix/planter-nonstop-complain

rivon0507 committed:
Convert the idleAtJob boolean to a JobStatus enum

To differentiate being idle because of efficiency and being idle because of problems.

rivon0507 committed:
Deserialization migration

rivon0507 committed:
Merge branch 'version/main' into fix/planter-nonstop-complain

rivon0507 commented:
So now we have a jobStatus (IDLE, WORKING, STUCK) enum instead of a idleAtJob boolean.

Raycoms commented on src/main/java/com/minecolonies/api/colony/jobs/IJob.java:
```diff
@@ -123,7 +124,7 @@ default void onLevelUp()
     default void initEntityValues(AbstractEntityCitizen citizen)
     {
-        citizen.getCitizenData().setIdleAtJob(false);
+        citizen.getCitizenData().setJobStatus(JobStatus.IDLE);
```
Initially this was WORK, right

Raycoms replied:
But I guess IDLE is fine

Raycoms commented on src/main/java/com/minecolonies/core/colony/CitizenData.java:
```diff
@@ -1491,7 +1492,15 @@ public void deserializeNBT(final CompoundTag nbtTagCompound)
+        final String jobStateString = nbtTagCompound.getString(TAG_JOB_STATUS);
+        if (!jobStateString.isEmpty())
+        {
+            this.jobStatus = JobStatus.valueOf(jobStateString);
```
I'd go by ordinal for this

Raycoms replied:
Stil l missing the ordinal change here

rivon0507 replied:
Yeah, sorry I missed the comment. I'm working on it :)

rivon0507 replied:
Done

rivon0507 committed:
Merge branch 'version/main' into fix/planter-nonstop-complain

rivon0507 committed:
Merge branch 'version/main' into fix/planter-nonstop-complain

rivon0507 committed:
Update job status serialization to use ordinal values

rivon0507 committed:
Merge remote-tracking branch 'origin/fix/planter-nonstop-complain' into fix/planter-nonstop-complain

rivon0507 committed:
Merge branch 'version/main' into fix/planter-nonstop-complain

Raycoms commented on src/main/java/com/minecolonies/core/colony/CitizenData.java:
```diff
@@ -233,9 +234,9 @@ public class CitizenData implements ICitizenData
+     * The current status of the citizen's working (?)
      */
-    private boolean idle;
+    private JobStatus jobStatus;
```
oh, we can just initialize this here with idle, so that it can't be null. simplifies reading then.

rivon0507 committed:
Initialize jobStatus to IDLE to simplify serialization logic
